lorentz.py

```diff
@@ -148,26 +151,37 @@ def __init__(self, pairwise_matrix, sample_size=10):
         self.pairwise_matrix = pairwise_matrix
         self.n_items = len(pairwise_matrix)
         self.sample_size = sample_size
+        self.arange = np.arange(0, self.n_items)
 
     def __len__(self):
         return self.n_items
 
     def __getitem__(self, i):
-        I = torch.Tensor([i]).squeeze().long()
-        while True:
-            j = random.randint(0, self.n_items - 1)
-            if j != i:
+        I = torch.Tensor([i + 1]).squeeze().long()
+        has_child = False
+        arange = np.random.permutation(self.arange)
+        for j in arange:
+            if self.pairwise_matrix[i, j] > 0:  # assuming no self loop
+                has_child = True
                 break
+        if not has_child:  # if no child go for parent
+            for j in arange:
+                if self.pairwise_matrix[j, i] > 0:  # assuming no disconneted nodes
+                    break
         min = self.pairwise_matrix[i, j]
-        indices = [
-            index
-            for index, is_less in enumerate(self.pairwise_matrix[i] < min)
-            if is_less
-        ][: self.sample_size]
-        # offset indices by 1 and pad with 0
+        arange = np.random.permutation(self.arange)
+        if has_child:
+            indices = [
+                x for x in arange if i != x and self.pairwise_matrix[i, x] < min
+            ][: self.sample_size]
+        else:
+            indices = [
+                x for x in arange if i != x and self.pairwise_matrix[x, i] < min
+            ][: self.sample_size]
         Ks = ([i + 1 for i in [j] + indices] + [0] * self.sample_size)[
```

Comment:
j does not need a +1?

Reply:
I had thought that's what I did wrong.

Reply:
Yup there's a small mistake. wait!

Reply:
`[i + 1 for i in [j] + indices]`
I don't see any problem in this,
as there's 'i + 1` it'll do the j+1